### PR TITLE
Add methods for channels from file descriptors

### PR DESCRIPTION
**Pull request by hcbdfinity**
Expose methods to allow creating server and client channels from existing
file descriptors.

### CODE REVIEW

File: src/channel.rs [hunk 1]
```diff
@@ -442,6 +442,23 @@ impl ChannelBuilder {
         let channel =
             unsafe { grpc_sys::grpc_insecure_channel_create(addr_ptr, args.args, ptr::null_mut()) };
 
+        self.connect_from_raw_channel(channel)
+    }
+
+    /// Build an insecure [`Channel`] taking over an established connection from
+    /// a file descriptor.
+    pub fn connect_from_fd(mut self, target: &str, fd: ::std::os::raw::c_int) -> Channel {
```
Comment (BusyJay) [inline src/channel.rs:450]: It should be `unsafe` and unix only.

File: src/channel.rs [hunk 2]
```diff
@@ -442,6 +442,23 @@ impl ChannelBuilder {
...
+        let target = CString::new(target).unwrap();
+        let target_ptr = target.as_ptr();
+        let channel =
+            unsafe { grpc_sys::grpc_insecure_channel_create_from_fd(target_ptr, fd, args.args) };
+
+        self.connect_from_raw_channel(channel)
+    }
+
+    /// Build a channel from a raw underlying core grpc channel
+    pub fn connect_from_raw_channel(self, channel: *mut grpc_channel) -> Channel {
```
Comment (BusyJay) [inline src/channel.rs:461]: Unnecessary function.

File: src/server.rs [hunk 3]
```diff
@@ -561,6 +561,18 @@ impl Server {
     pub fn bind_addrs(&self) -> impl ExactSizeIterator<Item = (&String, u16)> {
         self.core.binders.iter().map(|b| (&b.host, b.port))
     }
+
+    /// Add an rpc channel for an established connection represented as a file
+    /// descriptor.
+    pub fn add_insecure_channel_from_fd(&self, fd: ::std::os::raw::c_int) {
```
Comment (BusyJay) [inline src/server.rs:567]: Should be unsafe and unix only.

File: src/channel.rs (hunk 1 above)
Comment (BusyJay) [reply src/channel.rs:450]: Should explain what's target and who is responsible to open and close fd.

Comment (hcbdfinity) [reply src/channel.rs:450]: thanks -- will amend doc, but how do I mark it "posix-only"?

File: src/channel.rs (hunk 2 above)
Comment (hcbdfinity) [reply src/channel.rs:461]: no, this function is quite useful actually -- it allows binding custom rpc channels created by other means (e.g. third party custom transport implementations), maybe this warrants more documentation? also, might need marking as unsafe

**Comment by hcbdfinity** [comment]
Thanks for review, amended documentation and "unsafe" annotations -- please advice on how I could mark the function in question as "posix-only", I actually do not know how to do that besides adding a comment.

File: src/channel.rs (hunk 2 above)
Comment (BusyJay) [reply src/channel.rs]: Can you give more details about the use case? Is it a real use case?

File: src/channel.rs (hunk 1 above)
Comment (BusyJay) [reply src/channel.rs]: `#[cfg(unix)]`

File: src/channel.rs (hunk 2 above)
Comment (hcbdfinity) [reply src/channel.rs]: Might be a 'fringe' use case, although it is 'real' for me -- you basically always need that when you intend to use anything besides 'plain' sockets for transport.

File: src/channel.rs (hunk 1 above)
Comment (hcbdfinity) [reply src/channel.rs]: thanks, added as attribute

File: src/channel.rs [hunk 4]
```diff
@@ -442,6 +442,30 @@ impl ChannelBuilder {
...
+        let channel = grpc_sys::grpc_insecure_channel_create_from_fd(target_ptr, fd, args.args);
+
+        self.connect_from_raw_channel(channel)
+    }
+
+    /// Build a channel from a raw underlying core grpc channel. This allows to
+    /// establish an RPC connection using custom channel implementations.
+    /// Takes over ownership of the given channel object and will close it
+    /// when the connection is closed.
+    pub unsafe fn connect_from_raw_channel(self, channel: *mut grpc_channel) -> Channel {
```
Comment (BusyJay) [inline src/channel.rs:468]: You can mark `Channel::new` as public and unsafe instead.

File: src/server.rs [hunk 5]
```diff
@@ -561,6 +561,13 @@ impl Server {
     pub fn bind_addrs(&self) -> impl ExactSizeIterator<Item = (&String, u16)> {
         self.core.binders.iter().map(|b| (&b.host, b.port))
     }
+
+    /// Add an rpc channel for an established connection represented as a file
+    /// descriptor.
```
Comment (BusyJay) [inline src/server.rs:566]: Missing docs for `fd`.

**Comment by BusyJay** [comment]
You need to signoff your commit to get DCO pass. The [Details](https://github.com/probot/dco#how-it-works) link has instructions about how to do it.

File: src/channel.rs (hunk 4 above)
Comment (hcbdfinity) [reply src/channel.rs]: > You can mark `Channel::new` as public and unsafe instead.

sure, also fine with that solution -- changed

File: src/channel.rs [hunk 6]
```diff
@@ -548,7 +570,11 @@ unsafe impl Send for Channel {}
 unsafe impl Sync for Channel {}
 
 impl Channel {
-    fn new(cq: CompletionQueue, env: Arc<Environment>, channel: *mut grpc_channel) -> Channel {
+    pub unsafe fn new(
```
Comment (BusyJay) [inline src/channel.rs:573]: Doc is needed.

**Review by BusyJay — COMMENTED** [review]
Rest LGTM.

**Review by BusyJay — APPROVED** [review]
Thanks for your contributions! Please fix CI.

**Comment by hcbdfinity** [comment]
> Thanks for your contributions! Please fix CI.

Thanks for review and sorry for slight delay fixing issues -- all green now, let me know if anything else is required to proceed.

**Comment by hcbdfinity** [comment]
Gentle ping, anything that is required still for review?

**Review by BusyJay — APPROVED** [review]

**Comment by BusyJay** [comment]
@hunterlxt PTAL

**Comment by yiwu-arbug** [comment]
> Most of reviewers are on National holiday, I'm afraid it may not be merged till Friday.
> 
> /cc @yiwu-arbug can you take a look?

@BusyJay sorry, I didn't notice your comment :/

**Comment by hunterlxt** [comment]
LGTM, but maybe it needs an unit test to verify that this api works

**Review by hicqu — APPROVED** [review]

**Comment by BusyJay** [comment]
The method is only calling the unsafe one, I'm OK without a ut.

**Review by hunterlxt — APPROVED** [review]
LGTM

**Pull request merged**